### DISCUSSION TITLE
鼓励使用 auto 来自动推断初始化变量的类型吗?

### DISCUSSION THREAD

**Discussion by ShenMian** [BoomingTech/Piccolo#128] [Q&A] [answered]
请问该项目鼓励使用 `auto` 来自动推断初始化变量的类型吗? 例如:
```cpp
std::shared_ptr<PilotRenderer> render = m_engine_runtime->getRender();
```
改为
```cpp
auto render = m_engine_runtime->getRender();
```
因为目前没有注意到该项目尝试使用 `auto`, 所以想问一下新的代码可不可以(或应不应该)使用 `auto` 来自动推断初始化变量的类型.

**Comment by ambiguouStr** [comment]
It seems that the `.clang-tidy` file in the root path of this project is composed on purpose, 
and `modernize-use-auto` could not be found in it. 

Maybe we could use modern c++ features & attributes under the guidance of `.clang-tidy` 
and formate codes with `.clang-format`.

**Comment by Ol6rin** [comment, marked as answer]
过度使用auto会严重影响代码的可读性，提高维护成本，代码看到后面容易忘记之前的变量是什么类型，本项目严格限制auto的使用。
仅以下三种情况允许使用：
1. 定义lambda表达式；
2. 定义容器的迭代器变量；
3. 某些情况下为了避免非常冗长的写法。（如基于范围的for循环遍历对象的声明）
第3种情况需要注意引用的使用，避免不必要的内存拷贝。

后面我们会整理项目的编码规范，放在wiki页面中。
